Core/HW/StereoStretcher.cpp

```diff
@@ -0,0 +1,83 @@
+#include <algorithm>
+#include <numeric>
+#include "base/basictypes.h"
+#include "base/timeutil.h"
+#include "StereoStretcher.h"
+#include "Globals.h"
+
+StereoStretcher::StereoStretcher() {
+	stretch_.SetOutputSampleRate(48000);
+	mixed_samples_ = 0;
+	last_now_ = real_time_now();
+}
+
+unsigned int StereoStretcher::Mix(s16 *buffer, unsigned int numSamples, bool consider_framelimit, int sampleRate) {
+	mutex_.lock();
+	int remaining_size = numSamples * 2;
+	while (remaining_size > 0 && !queue.empty()) {
+		if (!queue.front().size()) {
+			queue.pop_front();
+		} else if (queue.front().size() <= remaining_size) {
+			memcpy(buffer, queue.front().data(), queue.front().size() * sizeof(s16));
+			buffer += queue.front().size();
+			remaining_size -= (int)queue.front().size();
+			queue.pop_front();
+		} else {
+			memcpy(buffer, queue.front().data(), remaining_size * sizeof(s16));
+			buffer += remaining_size;
+			queue.front().erase(queue.front().begin(), queue.front().begin() + remaining_size);
+			remaining_size = 0;
+		}
+	}
+	if (remaining_size > 0) {
+		memset(buffer, 0, remaining_size * sizeof(s16));
+	}
+	mutex_.unlock();
+	mixed_samples_ += numSamples;
+	return numSamples - remaining_size * 2;
+}
+
+void StereoStretcher::PushSamples(const s32 * samples, unsigned int num_samples) {
+	stats_.lastPushSize = num_samples;
+	s16 buffer[16384];
+	while (num_samples > 0) {
+		int mono_count = std::min(num_samples * 2, (unsigned int)ARRAY_SIZE(buffer));
+		for (int i = 0; i < mono_count; i++) {
+			buffer[i] = clamp_s16(samples[i]);
+		}
+		stretch_.AddSamples(buffer, mono_count / 2);
+		num_samples -= mono_count / 2;
+	}
+	
+	// Only process when we have a substantial amount of samples.
```

Comment:
So I guess this adds more of a delay to audio compared to video, etc. right?

-[Unknown]